src/main/java/org/mockito/internal/MockitoCore.java

```diff
@@ -60,11 +63,32 @@ public boolean isTypeMockable(Class<?> typeToMock) {
         }
         MockSettingsImpl impl = MockSettingsImpl.class.cast(settings);
         MockCreationSettings<T> creationSettings = impl.build(typeToMock);
+        checkDoNotMockAnnotation(creationSettings.getTypeToMock(), creationSettings);
         T mock = createMock(creationSettings);
         mockingProgress().mockingStarted(mock, creationSettings);
         return mock;
     }
 
+    private void checkDoNotMockAnnotation(Class<?> typeToMock,
+        MockCreationSettings<?> creationSettings) {
+        ArrayList<Class<?>> unmockableTypes = new ArrayList<>(
+            creationSettings.getExtraInterfaces());
+
+        Class<?> mockTypeToCheck = typeToMock;
+        while (mockTypeToCheck != null && mockTypeToCheck != Object.class) {
+            unmockableTypes.add(mockTypeToCheck);
+            Class<?>[] interfaces = mockTypeToCheck.getInterfaces();
+            if (interfaces != null) {
+                unmockableTypes.addAll(Arrays.asList(interfaces));
+            }
+            mockTypeToCheck = mockTypeToCheck.getSuperclass();
+        }
+
+        if (unmockableTypes.stream().anyMatch(clazz -> clazz.isAnnotationPresent(DoNotMock.class))) {
```

Comment:
Do we want to honor this specific annotation (Mockito API) or _any_ annotation with name "DoNotMock".

The latter seems more useful. For example: a library owner can mark types with any "DoNotMock" annotation without adding compile dependency on mockito-core.

Reply:
I would like to keep this for now and open it up later to match on name. If we would implement that now, I would break all users of ErrorProne that rely on `@DoNotMock` but have grandfathered mock violations.

If we can gather feedback first, we can see adoption patterns and change it to match on the name later. WDYT?

Reply:
How about a happy medium so that we don't need to revisit this in the future (and cause breaking changes :) - we can match for "org.mockito.DoNotMock" by String. This should be backwards compatible for G. and other companies / projects. Thoughts?

Reply:
Fixed